### PR TITLE
Version 2 for Raw Block Devices

### PR DESCRIPTION
**Pull request by erinboyd**

### CODE REVIEW

**Comment by erinboyd** [comment]
@thockin @saad-ali please assign this to me

**Comment by erinboyd** [comment]
This PR is based on  #805  which we can't open because of github issues

**Comment by erinboyd** [comment]
@jsafrane @msau42 please one final review :crossed_fingers:

File: contributors/design-proposals/raw-block-pv.md [hunk 1]
```diff
@@ -0,0 +1,739 @@
+# Raw Block Consumption in Kubernetes
```
Comment (msau42) [inline contributors/design-proposals/raw-block-pv.md]: You can remove this file

File: contributors/design-proposals/storage/raw-block-pv.md [hunk 2]
```diff
@@ -0,0 +1,739 @@
...
+devices = append(devices, kubecontainer.DeviceInfo{PathOnHost: path, PathInContainer: path, Permissions: "XXX"}) 
+
+The implemenation plan would be to rename the current makeDevices to makeGPUDevices and create a separate function to add the raw block devices to the option array to be passed to the container runtime. This would iterate on the paths passed in for the pod/container.
+
+Since the future of this in Kubernetes for GPUs and other plug-able devices is migrating to a device plugin architecture, there are 
+still differentiating components of storage that are enough to not to enforce alignment to their convention. Two factors when
+considering the usage of device plugins center around discoverability and topology of devices. Since neither of these are requirements
+for using raw block devices, the legacy method of populating the devices and appending it to the device array is sufficient.
+
+# Implementation Plan, Features & Milestones
```
Comment (msau42) [inline contributors/design-proposals/storage/raw-block-pv.md]: This should move to the bottom

File: contributors/design-proposals/storage/raw-block-pv.md [hunk 3]
```diff
@@ -0,0 +1,739 @@
...
+
+                Milestone 1: Dynamically provisioned PVs to dynamically allocated devices
+
+                Milestone 2: Privileged container concerns    
+		
+		Milestone 3: Flex volume update
+		
+Other considerations:
+
+               Reference volume driver change(Attach/Detach logic) for pre-provisioned PVs
```
Comment (msau42) [inline contributors/design-proposals/storage/raw-block-pv.md]: What are these two items?

File: contributors/design-proposals/storage/raw-block-pv.md [hunk 4]
```diff
@@ -0,0 +1,739 @@
...
+
+               Reference volume driver change(Provision/Delete logic) for dynamic provisioning
+
+# Plugin interface changes
```
Comment (msau42) [inline contributors/design-proposals/storage/raw-block-pv.md]: Mention here that volume plugins have to implement these new interfaces to support block.

File: contributors/design-proposals/storage/raw-block-pv.md [hunk 5]
```diff
@@ -0,0 +1,739 @@
...
+## New BlockVolume interface proposed design
+
+```
+// BlockVolume represents a directory used by pods or hosts on a node.
```
Comment (msau42) [inline contributors/design-proposals/storage/raw-block-pv.md]: directory?

File: contributors/design-proposals/storage/raw-block-pv.md [hunk 6]
```diff
@@ -0,0 +1,739 @@
...
+```
+
+## New BlockVolumePlugin interface proposed design
+
+```
+// BlockVolumePlugin is an extend interface of VolumePlugin and is used for block volumes support.
+type BlockVolumePlugin interface {
+	VolumePlugin
+	// NewBlockVolumeMapper creates a new volume.BlockVolumeMapper from an API specification.
+	// Ownership of the spec pointer in *not* transferred.
```
Comment (msau42) [inline contributors/design-proposals/storage/raw-block-pv.md]: What do you mean by ownership, and not transferred?

File: contributors/design-proposals/storage/raw-block-pv.md [hunk 7]
```diff
@@ -0,0 +1,739 @@
...
+
+```
+// BlockVolumeMapper interface provides methods to set up/map the volume.
+type BlockVolumeMapper interface {
+	BlockVolume
+	// SetUpDevice prepares the volume to a self-determined directory path,
+	// which may or may not exist yet and returns combination of physical
+	// device path of a block volume and error.
+	// If the plugin is non-attachable, it should prepare the device
+	// in /dev/ (or where appropriate) and return device path.
```
Comment (msau42) [inline contributors/design-proposals/storage/raw-block-pv.md]: Is there any requirement here that the device path is unique?

**Comment by msau42** [comment]
@kubernetes/sig-storage-pr-reviews

File: contributors/design-proposals/storage/raw-block-pv.md (hunk 6 above)
Comment (mtanino) [reply contributors/design-proposals/storage/raw-block-pv.md]: This comment is just copy and paste from VolumePlugin. will remove this line.

File: contributors/design-proposals/storage/raw-block-pv.md (hunk 4 above)
Comment (mtanino) [reply contributors/design-proposals/storage/raw-block-pv.md]: will do.

File: contributors/design-proposals/storage/raw-block-pv.md (hunk 5 above)
Comment (mtanino) [reply contributors/design-proposals/storage/raw-block-pv.md]:
```
BlockVolume interface provides methods to generate global map path and pod device map path.
```
is enough I suppose.

File: contributors/design-proposals/storage/raw-block-pv.md (hunk 7 above)
Comment (mtanino) [reply contributors/design-proposals/storage/raw-block-pv.md]: Currently there is no requirement for the device path.

**Comment by erinboyd** [comment]
[1] Block Volume Support API Change
VolumeMode and VolumeDevice api
https://github.com/kubernetes/kubernetes/pull/50457

[2] VolumeMode PV-PVC Binding change
WIP: VolumeMode binding logic update #53385
https://github.com/kubernetes/kubernetes/pull/53385

[3] Block volumes Support: CRI, volumemanager and operationexecutor changes
https://github.com/kubernetes/kubernetes/pull/51494

[4] Initial plugin changes
Block volumes Support: FC plugin update #51493
https://github.com/kubernetes/kubernetes/pull/51493

**Comment by fabiand** [comment]
Is this actually intended to land in 1.9?

**Comment by mtanino** [comment]
@fabiand @erinboyd 
Yes, we are actively contributing to this feature.

**Comment by fabiand** [comment]
@mtanino cool. thanks

**Comment by erinboyd** [comment]
@msau42 @jsafrane please review

**Comment by jsafrane** [comment]
lgtm. @msau42, WDYT?

**Comment by jsafrane** [comment]
/test all
"Required Github CI test is not green: pull-community-verify"

**Comment by mtanino** [comment]
## Current ongoing proposals

### API, Runtime, VolumeManager
[1-1] Block Volume Support API Change
kubernetes/kubernetes#50457

[1-2] VolumeMode PV-PVC Binding change
kubernetes/kubernetes#53385

[1-3] Block volumes Support: CRI, volumemanager and operationexecutor changes
kubernetes/kubernetes#51494

[1-4] Block volume: Command line printer update
kubernetes/kubernetes#55112

### Plugin
[2-1] Block volumes Support: FC plugin update #51493
kubernetes/kubernetes#51493

[2-2] Block volumes Support: iSCSI plugin update
kubernetes/kubernetes#54752

### external-provisioner
[3-1] [iscsi-targetd] block volume support
kubernetes-incubator/external-storage#440

**Pull request merged**